myia/grad.py

```diff
@@ -0,0 +1,284 @@
+        return self.tagged_graphs[graph]
+
+    def process_graph_forward(self, graph):
+        """Create the forward graph."""
+        if graph in self.done_fw:
+            return
+
+        if graph not in self.tagged_graphs:
+            self.scaffold_graph(graph)
+
+        tgraph = self.tagged_graphs[graph]
+        bgraph = self.backpropagator_graphs[graph]
+
+        # Return (↑graph.output, ♢graph). The first element is given
+        # by the `phi` method.
+        tgraph.output = Apply([
+            mktuple,
+            self.phi(graph.output),
+            Constant(bgraph)
+        ], tgraph)
+
+        self.done_fw.add(graph)
+
+    def process_all_graphs_forward(self, root):
+        """Create the forward graph for all graphs starting from root."""
+        self.todo_fw.add(root)
+        while self.todo_fw:
+            g = self.todo_fw.pop()
+            self.process_graph_forward(g)
+
+    def process_graph_backward(self, graph):
+        """Create the backward graph."""
+        bgraph = self.backpropagator_graphs[graph]
+
+        # Return ((∇fv1, ∇fv2, ...), ∇arg1, ∇arg2, ...)
+        # Where ∇x is given by `rho(x, graph)`
+        bgraph.output = Apply([
+            mktuple,
+            Apply([
+                mktuple,
+                *[self.rho(p, graph)
+                  for p in self.fv_order[graph]]
+            ], bgraph),
+            *[self.rho(p, graph)
+              for p in graph.parameters]
+        ], bgraph)
+
+    def process_all_graphs_backward(self):
+        """Create the backward graph for all graphs.
+
+        This method has no argument: it simply processes all graphs for which we
+        created a forward graph.
+        """
+        for g in self.todo_bw:
+            self.process_graph_backward(g)
+
+    def phi(self, node):
+        """Compute equivalent node in forward graph."""
+        if node in self.tagged_nodes:
+            return self.tagged_nodes[node]
+
+        tg = node.graph and self.tagged_graphs[node.graph]
+
+        if is_graph(node):
+            # We will have to process this graph too.
+            tagged, bprop = self.scaffold_graph(node.value)
+            self.graph_to_ct[node.value].add(node)
+        elif is_constant(node):
+            # Note that this application will have its graph set to None, which
+            # makes sense since it's basically a constant expression.
+            tagged, bprop = Apply([J, node], tg), None
+        else:
+            # a = f(x, y) -> ↑a, ♢a = ↑f(↑x, ↑y)
+            tagged_args = [self.phi(n) for n in node.inputs]
+            app = Apply(tagged_args, tg)
+            # ↑a (the first element)
+            tagged = Apply([index, app, Constant(0)], tg)
+            # ♢a (the second element)
+            # Note that ♢a is not part of the forward graph, however,
+            # it will be a free variable of the backpropagator graph.
+            bprop = Apply([index, app, Constant(1)], tg)
```

Comment:
Do I understand this code correctly that primitives and functions are treated equally? Should we treat them equally? It seems like it's best to assume that the adjoints of primitives should always be inlined in the backpropagator, and perform this inlining immediately. I imagine this simplifies the graph a lot. So instead of this:
```python
def pow(x, y):
    def dpow(dz):
        return y * x ** (y - 1) * dz, x ** y * log(x) * dz
    return x ** y, dpow

def f(x):
    y, dpow = pow(x, 2)
    def df(dy):
        dx, _ = dpow(dy)
        return dx
    return y
```
can we generate this?
```python
def f(x):
    y = x ** 2
    def df(dy):
        return 2 * x * dy
    return y
```
After all, we know that primitives never have free variables, and we don't want to incur function call overhead each time a primitive's adjoint is used.

Reply:
We can optimize the graph into that. I don't see a good reason to do it immediately: I think that's unnecessary clutter and coupling with the optimization framework, and the less `Grad` does, the easier it will be to debug.